codec/enforceproto/extraneous_fields.go

```diff
@@ -0,0 +1,200 @@
+// compare its fields against those in msg. The error returned if non-nil will contain
+// the listing of the extraneous fields by tagNumber and wireType.
+func CheckExtraneousFieldsProto(b []byte, msg proto.Message) error {
+	pr := msg.ProtoReflect()
+	fields := pr.Descriptor().Fields()
+	unmatchedFields := make([]*tagAndWireType, 0, 100)
+	for len(b) > 0 {
+		tagNum, wireType, n := protowire.ConsumeField(b)
+		if n < 0 {
+			return errors.New("tagLen is invalid")
+		}
+		fd := fields.ByNumber(tagNum)
+		if fd == nil {
+			if tagNum&bit11Reserved == 0 {
+				// The tag is not reserved so report it.
+				unmatchedFields = append(unmatchedFields, &tagAndWireType{
+					tagNum:   tagNum,
+					wireType: wireType,
+				})
+			}
+		}
+
+		// TODO: Perhaps also assert on fd.Kind() if it doesn't match
+		// the exact type that we expect. However, gogoproto.Unmarshal will
+		// perform that check when invoked, but protobuf/proto unfortunately doesn't.
+		// See https://github.com/golang/protobuf/issues/1157
+		b = b[n:]
+	}
+	if len(unmatchedFields) == 0 {
+		// No extraneous fields so return.
+		return nil
+	}
+
+	// Otherwise report the unwarranted numbers.
+	return &UnmatchedTagsError{unmatchedFields}
+}
+
+type tagAndProtoType struct {
+	tagNum    protowire.Number
+	protoType string
+}
+
+func (tpt *tagAndProtoType) String() string {
+	return fmt.Sprintf("{tagNumber: %d, wireType:%s}", tpt.tagNum, tpt.protoType)
+}
+
+var typeToIndexFieldsMu sync.Mutex
+var typeToIndexFieldsForType = make(map[reflect.Type]map[protowire.Number]*tagAndProtoType)
+
+func indexFieldsForType(typ reflect.Type) (map[protowire.Number]*tagAndProtoType, error) {
+	typeToIndexFieldsMu.Lock()
+	defer typeToIndexFieldsMu.Unlock()
+
+	index, ok := typeToIndexFieldsForType[typ]
+	if ok {
+		return index, nil
+	}
+
+	index = make(map[protowire.Number]*tagAndProtoType)
+	for i, n := 0, typ.NumField(); i < n; i++ {
+		field := typ.Field(i)
+		protoTag, ok := field.Tag.Lookup("protobuf")
+		if !ok {
+			continue
+		}
+		byCommaSplits := strings.Split(protoTag, ",")
+		if len(byCommaSplits) == 0 {
+			// Strange, how can a protobuf tag not be populated?
+			continue
+		}
+		wireType, protoFieldNumStr := byCommaSplits[0], byCommaSplits[1]
+		protoField, err := strconv.ParseUint(protoFieldNumStr, 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf("invalid protoFieldNum in %q", protoFieldNumStr)
+		}
+		tagNum := protowire.Number(protoField)
+		index[tagNum] = &tagAndProtoType{
+			tagNum:    tagNum,
+			protoType: wireType,
+		}
+	}
+
+	// By this point we should memoize the index.
+	typeToIndexFieldsForType[typ] = index
+
+	return index, nil
+}
+
+// CheckExtraneousFieldsGogoProto walks through the protobuf serialized bytes in b, and tries to
+// compare its fields against those in msg. The error returned if non-nil will contain
+// the listing of the extraneous fields by tagNumber and wireType.
+func CheckExtraneousFieldsGogoProto(b []byte, msg gogoproto.Message) error {
```

Comment:
The logic for this looks right for a single message as far as I can tell, but it doesn't seem like it will examine nested messages and I don't see nested message handled in tests.

Reply:
I've added tests for nested messages too, PTAL.